src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1255,54 +1289,63 @@
-                                    :msg "compiling accumulate with join filter node"}]
-             (compile-expr id
-                           (compile-join-filter (:join-filter-expressions beta-node)
-                                                (:join-filter-join-bindings beta-node)
-                                                (:env beta-node))))
-           (:result-binding beta-node)
-           children
-           join-bindings
-           (:new-bindings beta-node))
+          (let [join-filter-expr (compile-join-filter (:join-filter-expressions beta-node)
+                                                      (:join-filter-join-bindings beta-node)
+                                                      (:env beta-node))]
+            (with-meta
+              (eng/->AccumulateWithJoinFilterNode
+               id
+               ;; Create an accumulator structure for use when examining the node or the tokens
+               ;; it produces.
+               {:accumulator (:accumulator beta-node)
+                ;; Include the original filter expressions in the constraints for inspection tooling.
+                :from (update-in condition [:constraints]
+                                 into (-> beta-node :join-filter-expressions :constraints))}
+               compiled-accum
+               (binding [*compile-ctx* {:condition condition
+                                        :join-filter-expressions (:join-filter-expressions beta-node)
+                                        :env (:env beta-node)
+                                        :msg "compiling accumulate with join filter node"}]
+                 (compile-expr id
+                               join-filter-expr))
+               (:result-binding beta-node)
+               children
+               join-bindings
+               (:new-bindings beta-node))
+              {:accum-expr accum-expr
+               :join-filter-expr join-filter-expr}))
 
           ;; All unification is based on equality, so just use the simple accumulate node.
-          (eng/->AccumulateNode
-           id
-           ;; Create an accumulator structure for use when examining the node or the tokens
-           ;; it produces.
-           {:accumulator (:accumulator beta-node)
-            :from condition}
-           compiled-accum
-           (:result-binding beta-node)
-           children
-           join-bindings
-           (:new-bindings beta-node))))
+          (with-meta
+            (eng/->AccumulateNode
+             id
+             ;; Create an accumulator structure for use when examining the node or the tokens
+             ;; it produces.
+             {:accumulator (:accumulator beta-node)
+              :from condition}
+             compiled-accum
+             (:result-binding beta-node)
+             children
+             join-bindings
+             (:new-bindings beta-node))
+            {:accum-expr accum-expr})))
```

Comment:
Should we make these items fields on the nodes themselves rather than using metadata? It seems like metadata just obscures things from someone who might be wanting to inspect or debug the nodes themselves, as opposed to just having it on the record.